### DISCUSSION TITLE
Hub for Godot Engine

### DISCUSSION THREAD

**Discussion by totchi-lagawi** [godotengine/godot-proposals#6030] [General Discussions]
I have a suggestion : a hub for the Engine, like Unity Hub, or the Epic Launcher (or Game Store, I don't know) for Unreal Engine. I think that can be really useful, for quickly update the Engine, install an other version... What do you think about that? Is this a good Idea? And excuse me for the bad english, I'm french, and I'm still learning...

**Comment by sammonius** [comment]
I don't really see the point. Godot projects don't need re-importing like Unity ones do, so why would you need multiple versions in the first place?

(I'm not a Godot developer, I was just scrolling by)

**Reply by Zireael07** [reply]
At least right now, GLES 3 fallback isn't a thing, so if your computer can't run Vulkan and you don't know how to set Godot 4 to GLES, you're ending with a very unfriendly message.

That's one reason. Another is that some people might not want to migrate to 4 yet or migrate only SOME projects and not all.

**Comment by theraot** [comment]
I love the lack of friction when getting a new version of Godot. It is quick: download, extract, done. It does not even require elevated privileges. It works well from a USB stick. You can have multiple versions (put them on separate folders).

So, you want a hub where you can get different versions of the engine? Here is the hub: https://downloads.tuxfamily.org/godotengine/ bookmark it and you are set.

I love that you don't have to run an installer, much less download and install the thing that actually downloads the installer for the engine.

Anyway, if you want to make sure you keep Godot up to date, and you want to get it bundled with all the export templates and stuff, you could also get it from Steam.

There are also launchers made by the community (using Godot, and fetching from the link I posted above, no less).

However I find it easy enough to have a folder. Uninstall? Delete the folder. Install? Download and extract. No fuss. Your operating system should have good UX for these operations. And you can handle the export template from Godot itself already.

**Reply by sammonius** [reply]
I think the main idea was that you have all projects available to all versions. If you're migrating a project, you just change a setting of what version that project opens in and you're set. It shouldn't change the folder system or make it harder to install.

**Reply by theraot** [reply]
@sammonius There is a list of projects for each major version of Godot. So, every Godot 3.x version will see the same list of projects.

There is no need to modify the projects (there is no setting of which version they require, and no need to move the projects or change their folder structure, and what installation are you talking about?). Granted, opening a project in an older version that lacks something the project needs will result in the game not working correctly.

The only big change is between major version (the current case is opening a Godot 3.x project in Godot 4).

Edit: To be clear, Godot 4 has a separate list of projects form Godot 3.x. So you would tell Godot 4 to scan the project, and upon opening it for the first time in Godot 4, it will offer you to convert it. Edit 2: It is a good idea to make a copy of the project anyway, since there is no going back. However, I reiterate that this is only because it is a change between major versions.

**Reply by theraot** [reply]
@sammonius Blow by blow, this what you do if you have a project in a version of Godot and you want it in a newer minor version.

- Assumption: you have a version of Godot and the project in it.
- Download the newer minor version of Godot
- Extract it
- Open it
- Open the project (since you get the same list of project, and you already had the project in the old version).

And if you want to go to the old version? Open it (assuming you didn't delete it), and open the project (it is the same list of projects).

If it were form Godot 3 to 4, it is different:

- Assumption: you have a version of Godot 3 and the project in it.
- Download the newer version of Godot 4
- Extract it
- Open it
- Tell Godot to scan the location of the project
- Open the project (once the location is scanned it should apear on the list).
- Tell Godot to convert the project (it asks)
- And then you have about 45 minutes to an hour of fixing errors, because the conversion is not perfect

And if you want to go to the old version? NOPE. So have a back up before the conversion.

**Reply by sammonius** [reply]
@theraot Sorry, my reply was poorly worded. What I meant was that having a hub would allow you to install a new version without having to open each project by folder again. It saves a lot of pain if you have many small projects and want to download a minor update.

**Reply by theraot** [reply]
@sammonius I do not understand how that is a pain, since - I reiterate - the have the same list of projects.

I can, for example, create a project in Godot 3.5.1 and then go to open Godot 3.2 and the project is in the list. Or viceversa.

And you know what? It is the same editor settings. I can - for example - go into Godot 3.2 and change the accent color of the editor theme, and then open Godot 3.5.2 RC1 and see the accent color changed. And yes, if you have configured Godot to automatically scan a folder (I don't), that works for that too.

**Reply by Zireael07** [reply]
What @theraot said. The only case where you would need the hub/launcher thingy is 3 -> 4. Upgrading to minor versions is as easy as unpacking a zip and opening Godot, the project list is the **same**

**Comment by dalexeev** [comment]
If I understand correctly, the suggestion of this discussion is to create a launcher (instead of Project Manager) that runs different Godot editor executables for different projects. For example, different minor versions or custom builds. This could be useful for those who maintain legacy projects that cannot be updated to the latest stable version (for one reason or another).

The good news is that it is possible to implement this unofficially, as there is a command line argument to run the editor bypassing the Project Manager, and the Godot editor supports [Self-contained mode](https://docs.godotengine.org/en/stable/tutorials/io/data_paths.html#doc-data-paths-self-contained-mode) (to isolate settings between different versions/builds). Also, this launcher can have the functionality of automatically downloading and unpacking binaries from downloads.tuxfamily.org when requested by the user.

Since this feature is not very popular (as far as I know), I think it's better to implement such a launcher unofficially first. And only if it becomes popular enough, we can consider replacing the Project Manager.

Related:

* #2329

**Comment by OverloadedTech** [comment]
According to the idea of the Unity Hub i think that implementing it also in Godot will be very useful to manage installations. Project can also ba managed in the editor version change
So yes it is a good idea

(personal opinion)

**Comment by dsnopek** [comment]
Some folks have already made Godot launchers that allow you to connect each project with a specific version of Godot. Here's two examples I just googled up:

- https://eumario.itch.io/godot-manager
- https://nathan-warden.itch.io/godot-launcher

... but I don't think they're the only ones.

In any case, I think it makes sense to have this exist outside of the main Godot project for now. If it turns out a significant number of developers start using a particular launcher, maybe one of them could become official-ish?

**Reply by Zireael07** [reply]
Thanks for finding those - I was like 90% something already exists but I couldn't find :P

**Reply by OverloadedTech** [reply]
Even if they are unofficial it looks like they work really good without problems
